<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

If the initial code cells error out, apply suggestions from [this article](https://levelup.gitconnected.com/how-to-fix-attributeerror-optionengine-object-has-no-attribute-execute-in-pandas-eb635fbb89e4) to fix, to successfully work on lab.

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Install sqlalchemy & psycopg2 - You only need to do this once!

# !pip install sqlalchemy
# !pip install psycopg2

In [2]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

Can you display the count of rows in the `stores` table of the Iowa liquor database?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    COUNT(*)
FROM 
    stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**:

1973 rows.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    store, name
FROM
    stores
ORDER BY 
    name
LIMIT 1
OFFSET 3
"""

pd.read_sql_query(sql, engine)

store                         name
0   3061  3061 Cub Foods / Sioux City

**Prompt 2 Answer**:

3061 Cub Foods / Sioux City.

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    category_name, btl_price, bottle_qty, vendor
FROM 
    sales
WHERE 
    category_name = 'IMPORTED VODKA'
ORDER BY 
    vendor
LIMIT 3
"""

pd.read_sql_query(sql, engine)

category_name btl_price  bottle_qty                     vendor
0  IMPORTED VODKA     $8.99          12  Adamba Imports Int'l Inc.
1  IMPORTED VODKA    $28.50           1      A Hardy / U.S.A. Ltd.
2  IMPORTED VODKA    $28.50           6      A Hardy / U.S.A. Ltd.

**Prompt 3 Answer**: 

8.99, 12 bottles.

28.50 6 bottles.

28.50 3 bottles.

### Prompt 4

What is the total revenue of liquor sold in February 2015?

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """
WITH day_totals AS (
    SELECT 
        SUM(total) as day_total, 
        date
    FROM 
        sales
    WHERE 
        date between '2015-02-01' and '2015-02-28' 
    GROUP BY 
        date
)

SELECT
    SUM(day_total) as month_total
FROM
    day_totals
"""

pd.read_sql_query(sql, engine)

month_total
0  21295350.55

In [27]:
# Using SQL subquery...

sql = """
SELECT
    SUM(day_total) as month_total
FROM
    (SELECT 
        SUM(total) as day_total, 
        date
    FROM 
        sales
    WHERE 
        date between '2015-02-01' and '2015-02-28' 
    GROUP BY 
        date
) as day_totals
"""

pd.read_sql_query(sql, engine)

month_total
0  21295350.55

**Prompt 4 Answer**:

21295350.55 dollars.

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [15]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    SUM(total) as revenue, SUM(bottle_qty) as bottles_sold, COUNT(vendor) as total_transactions, vendor
FROM 
    sales
WHERE 
    category_name = 'IMPORTED VODKA'
GROUP BY 
    vendor
ORDER BY
    total_transactions DESC
LIMIT 1

"""

pd.read_sql_query(sql, engine)

revenue  bottles_sold  total_transactions  \
0  7880365.88        403450               42338   

                             vendor  
0  Pernod Ricard USA/Austin Nichols

**Prompt 5 Answer**: 

Pernod Ricard USA/Austin Nichols.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT DISTINCT 
    description, (btl_price - state_btl_cost) AS Markup
FROM 
    sales
ORDER BY
    Markup DESC
LIMIT 10

"""

pd.read_sql_query(sql, engine)

description     markup
0                 Cedar Ridge Barrel Proof Bourbon  $2,900.00
1             Absolut w/ Zing Zang Bloody Mary Mix    $799.60
2                    Remy Martin Louis Xiii Cognac    $699.65
3     Macallan Sherry Cask 25 Year Old Scotch - HA    $281.70
4                           Johnnie Walker Odyssey    $249.32
5                        Johnnie Walker Odyssey HA    $249.32
6  Johnnie Walker Private Collection 2014 Smoke HA    $212.50
7  Johnnie Walker Private Collection 2015 Fruit HA    $212.50
8                   Macallan 25yr Anniversary Malt    $212.39
9                Macallan 25yr Anniversary Malt HA    $212.39

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    stores.name,
    AVG((btl_price - state_btl_cost) / state_btl_cost) * 100 AS avg_percent_markup
FROM 
    stores
JOIN 
    sales ON stores.store = sales.store
GROUP BY 
    stores.name
ORDER BY 
    avg_percent_markup DESC
LIMIT 1

"""

pd.read_sql_query(sql, engine)

name  avg_percent_markup
0  Wal-Mart 0841 / Tipton           52.547124

**Prompt 7 Answer**: 

Wal-Mart 0841 / Tipton,	52.547124.

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, as well as the names and addresses of those store.

What are the names of the five stores?

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    stores.store,
    stores.name,
    stores.store_address,
    AVG((sales.btl_price - sales.state_btl_cost) / sales.state_btl_cost) * 100 AS avg_percent_markup
FROM 
    stores
JOIN 
    sales ON stores.store = sales.store
GROUP BY 
    stores.name, stores.store, stores.store_address
ORDER BY 
    avg_percent_markup DESC
LIMIT 5;

"""

pd.read_sql_query(sql, engine)

store                        name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   
4   3660   Wal-Mart 2935 / Knoxville   

                                       store_address  avg_percent_markup  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...           52.547124  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...           52.149880  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...           51.890342  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....           51.509378  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...           51.459084

**Prompt 8 Answer**:

Wal-Mart 0841 / Tipton	

Wal-Mart 1546 / Iowa Falls

Wal-Mart 1683 / Shenandoah

Wal-Mart 3394 / Atlantic

Wal-Mart 2935 / Knoxville

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [12]:
# Use a SQL query to find the answer to the above prompt.

sql = """
WITH TopNine AS (
    SELECT 
        c.county, 
        SUM(s.bottle_qty) / c.population AS bottles_per_capita,
        SUM(s.bottle_qty * s.liter_size) / SUM(s.bottle_qty) AS avg_bottle_size
    FROM 
        sales s
    JOIN 
        counties c ON s.county = c.county
    GROUP BY 
        c.county
    ORDER BY
        bottles_per_capita DESC,
        avg_bottle_size DESC
    LIMIT 9
)

SELECT 
    county,
    avg_bottle_size,
    bottles_per_capita
FROM 
    TopNine
ORDER BY
    avg_bottle_size DESC
LIMIT 1;
"""

pd.read_sql_query(sql, engine)

county  avg_bottle_size  bottles_per_capita
0  Kossuth             1119                  11

In [26]:
# Using SQL subquery...

sql = """
SELECT
    county,
    avg_bottle_size,
    bottles_per_capita
FROM
    (SELECT 
        c.county, 
        SUM(s.bottle_qty) / c.population AS bottles_per_capita,
        SUM(s.bottle_qty * s.liter_size) / SUM(s.bottle_qty) AS avg_bottle_size
    FROM 
        sales s
    JOIN 
        counties c ON s.county = c.county
    GROUP BY 
        c.county
    ORDER BY
        bottles_per_capita DESC,
        avg_bottle_size DESC
    LIMIT 9
) as top_nine
ORDER BY
    avg_bottle_size DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

county  avg_bottle_size  bottles_per_capita
0  Kossuth             1119                  11

**Prompt 9 Answer**:

Kossuth. 

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [30]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
    COUNT(*) AS num_transactions_not_equal
FROM (
    SELECT 
        btl_price,
        bottle_qty,
        CAST(total AS MONEY) as money_total, 
        (btl_price * bottle_qty) AS calculated_total,
        CASE
            WHEN CAST(total AS MONEY) = (btl_price * bottle_qty) THEN 'Equal'
            ELSE 'Not Equal'
        END AS comparison_result
    FROM
        sales
) AS subquery
WHERE
    comparison_result = 'Not Equal';
"""

pd.read_sql_query(sql, engine)

num_transactions_not_equal
0                           0

**Prompt 10 Answer**:

The answer is 0.